src/journal-remote/journal-gatewayd.c

```diff
@@ -885,6 +888,139 @@ static int request_handler_machine(
+        int r;
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *json = NULL;
+
+        r = sd_json_build(
+                        &json,
+                        SD_JSON_BUILD_OBJECT(
+                                        SD_JSON_BUILD_PAIR("index", SD_JSON_BUILD_INTEGER(boot_display_index)),
+                                        SD_JSON_BUILD_PAIR_ID128("boot_id", boot.id),
+                                        SD_JSON_BUILD_PAIR("first_entry", SD_JSON_BUILD_UNSIGNED(boot.first_usec)),
+                                        SD_JSON_BUILD_PAIR("last_entry", SD_JSON_BUILD_UNSIGNED(boot.last_usec))));
+        if (r < 0)
+                return r;
+
+        r = sd_json_variant_dump(json, SD_JSON_FORMAT_SEQ, f, NULL);
+        if (r < 0)
+                return r;
+
+        return 0;
+}
+
+static ssize_t request_reader_boots(
+                void *cls,
+                uint64_t pos,
+                char *buf,
+                size_t max) {
+
+        RequestMeta *m = ASSERT_PTR(cls);
+        int r;
+        size_t n, k;
+        LogId boot;
+
+        assert(buf);
+        assert(max > 0);
+        assert(pos >= m->delta);
+
+        pos -= m->delta;
+
+        while (pos >= m->size) {
+                off_t sz;
+
+                /* We're seeking from tail (newest boot) so advance to older. */
+                r = discover_next_id(
+                                m->journal,
+                                LOG_BOOT_ID,
+                                /* boot_id */ SD_ID128_NULL,
+                                /* unit = */ NULL,
+                                m->previous_boot_id,
+                                /* advance_older = */ true,
+                                &boot);
+                if (r < 0) {
+                        log_error_errno(r, "Failed to advance boot index: %m");
+                        return MHD_CONTENT_READER_END_WITH_ERROR;
+                }
+                if (r == 0)
+                        return MHD_CONTENT_READER_END_OF_STREAM;
+
+                pos -= m->size;
+                m->delta += m->size;
+
+                r = request_meta_ensure_tmp(m);
+                if (r < 0) {
+                        log_error_errno(r, "Failed to create temporary file: %m");
+                        return MHD_CONTENT_READER_END_WITH_ERROR;
+                }
+
+                r = output_boot(m->tmp, boot, m->boot_index);
+                if (r < 0) {
+                        log_error_errno(r, "Failed to serialize boot: %m");
+                        return MHD_CONTENT_READER_END_WITH_ERROR;
+                }
+
+                sz = ftello(m->tmp);
+                if (sz < 0) {
+                        log_error_errno(errno, "Failed to retrieve file position: %m");
+                        return MHD_CONTENT_READER_END_WITH_ERROR;
+                }
+
+                m->size = (uint64_t) sz;
+
+                m->previous_boot_id = boot.id;
+                m->boot_index -= 1;
```

Comment:
Also here. Why `previous_boot_id` and `boot_index` need to be saved in Manager??

Reply:
Answer/question similar to the previous comment - but here to keep track of boot index between individual content reader calls.

Reply:
Right, thanks!